core/src/Streamly/Internal/Data/Serialize/TH/Bottom.hs

```diff
@@ -65,48 +74,122 @@ import Streamly.Internal.Data.Unbox.TH (DataCon(..))
 -- Config
 --------------------------------------------------------------------------------
 
--- | Config to control how the 'Serialize' instance is generated.
+-- XXX Need to make it (Maybe Inline) so that we can use Nothing as default.
+
+-- | Configuration to control how the 'Serialize' instance is generated. Use
+-- 'defaultConfig' and config setter functions to generate desired Config. For
+-- example:
+--
+-- >>> (inlineSize Inline) . (inlineSerialize Inlinable) defaultConfig
+--
 data Config =
     Config
-        { inlineSize :: Inline
-          -- ^ Inline value for 'size'. Default is Inline.
-        , inlineSerialize :: Inline
-          -- ^ Inline value for 'serialize'. Default is Inline.
-        , inlineDeserialize :: Inline
-          -- ^ Inline value for 'deserialize'. Default is Inline.
-        , constructorTagAsString :: Bool
-          -- ^ __Experimental__
-          --
-          -- If True, encode constructors using the constructor names as Latin-1
-          -- byte sequence.
-        , recordSyntaxWithHeader :: Bool
-          -- ^ __Experimental__
-          --
-          -- If True, encode the keys of the record as a header and then
-          -- serialize the data. Multiple constructors are not supported with
-          -- @recordSyntaxWithHeader@ enabled.
-          --
-          -- __Performance Notes:__
-          --
-          -- There is a constant regression proportional to
-          -- @sum (map length keyList) + length keyList@ where @keyList@ is the
-          -- list of keys of that record as strings.
-          --
-          -- As an example, @keyList@ of,
-          -- @
-          -- data RecordType = RecordType { field0 :: Int, field2 :: Char }
-          -- @
-          -- is @["field0", "field1"]@
+        { cfgInlineSize :: Inline
+        , cfgInlineSerialize :: Inline
+        , cfgInlineDeserialize :: Inline
+        , cfgConstructorTagAsString :: Bool
+        , cfgRecordSyntaxWithHeader :: Bool
         }
 
+-- | How should we inline the 'size' function? The default in 'defaultConfig'
+-- is 'Inline'. However, aggressive inlining can bloat the code and increase
+-- compilation times in when there are big functions and too many nesting
+-- levels so you can change it accordingly.
+inlineSize :: Inline -> Config -> Config
+inlineSize v cfg = cfg {cfgInlineSize = v}
+
+-- | How should we inline the 'serialize' function? See guidelines in
+-- 'inlineSize'.
+inlineSerialize :: Inline -> Config -> Config
+inlineSerialize v cfg = cfg {cfgInlineSerialize = v}
+
+-- | How should we inline the 'deserialize' function? See guidelines in
+-- 'inlineSize'.
+inlineDeserialize :: Inline -> Config -> Config
+inlineDeserialize v cfg = cfg {cfgInlineDeserialize = v}
+
+-- XXX We can identify if it is a sum type or a product type and use a single
```

Comment:
We need two different options.
```
data A = B {..} | C {..}
```
Encoding 'B', and 'C' as String and encoding the record syntax in a different way are 2 independent options.

Reply:
This case is currently unsupported though.

Reply:
I'm removing this `XXX` comment.